internal/pkg/runtime/engines/singularity/prepare_linux.go

```diff
@@ -720,6 +721,20 @@ func (e *EngineOperations) PrepareConfig(starterConfig *starter.Config) error {
 		starterConfig.SetCapabilities(capabilities.Ambient, e.EngineConfig.OciConfig.Process.Capabilities.Ambient)
 	}
 
+	// check if there is a controlling terminal
+	pgrp, _ := unix.IoctlGetInt(0, unix.TIOCGPGRP)
+	if pgrp == 0 {
+		pgrp, _ = unix.IoctlGetInt(1, unix.TIOCGPGRP)
+		if pgrp == 0 {
+			pgrp, _ = unix.IoctlGetInt(2, unix.TIOCGPGRP)
+		}
+	}
+	// cases we need to propagate signals to container process:
+	// - when pgrp == 0 because container won't run in a terminal
+	// - when process group is different from the process group controlling terminal
+	if pgrp == 0 || (pgrp > 0 && pgrp != syscall.Getpgrp()) {
```

Comment:
reading thru [ioctl_tty](http://man7.org/linux/man-pages/man2/ioctl_tty.2.html) and [tcgetpgrp](http://man7.org/linux/man-pages/man3/tcgetpgrp.3.html), the bit that stands out is:

>  and some value larger than 1 that is not presently a process group ID otherwise

is that the behavior of the system call or the GNU libc wrapper?

From those docs there seem to be three cases:

- If the file descriptor refers to the controlling terminal, and there's a foreground process associated with that terminal, the process group ID of that process is returned
- If the file descriptor refers to the controlling terminal, and there is no foreground process associated with that terminal, _some_ process group ID (associated with that terminal?) is returned
- if the file descriptor does not refer to the controlling terminal, -1 is returned

I can't find anything regarding 0.

If the behavior of the system call is different from the GNU libc wrapper, this should be documented so that other people don't have to go around chasing documentation to figure out if this is correct or not.

Reply:
This is syscall behaviour, in kernel source code 0 can be returned based on this code https://elixir.bootlin.com/linux/latest/source/kernel/pid.c#L416, I checked this is like that since a very long time. Now if there is any error with above code, the ioctl referenced value won't be affected and will follow the Go idiomatic way of a function returning an error where we should get zero default value